In [4]:
import apyori

In [5]:
help(apyori)

Help on module apyori:

NAME
    apyori - a simple implementation of Apriori algorithm by Python.

CLASSES
    builtins.object
        TransactionManager
    builtins.tuple(builtins.object)
        OrderedStatistic
        RelationRecord
        SupportRecord
    
    class OrderedStatistic(builtins.tuple)
     |  OrderedStatistic(items_base, items_add, confidence, lift)
     |  
     |  OrderedStatistic(items_base, items_add, confidence, lift)
     |  
     |  Method resolution order:
     |      OrderedStatistic
     |      builtins.tuple
     |      builtins.object
     |  
     |  Methods defined here:
     |  
     |  __getnewargs__(self)
     |      Return self as a plain tuple.  Used by copy and pickle.
     |  
     |  __repr__(self)
     |      Return a nicely formatted representation string
     |  
     |  _asdict(self)
     |      Return a new dict which maps field names to their values.
     |  
     |  _replace(self, /, **kwds)
     |      Return a new OrderedStatistic ob

In [6]:
tr = [['우유','계란','커피'],
     ['빵','우유','버터','휴지','라면'],
     ['맥주','땅콩'],
     ['계란','빵','우유','생수'],
     ['햇반','커피','간장','계란']]

In [9]:
import apyori as ap

result = ap.apriori(tr, min_support=0.3, \
                   min_confidence = 0.6)
rules = list(result) # 결과의 반환형식을 리스트로 꼭 바꿔주어야 한다. ***

In [11]:
for r in rules: 
    print(r, "\n") # 중간 공백 만들어줌

# 큰 세 덩어리로 이루어져 있다. 

RelationRecord(items=frozenset({'계란'}), support=0.6, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'계란'}), confidence=0.6, lift=1.0)]) 

RelationRecord(items=frozenset({'우유'}), support=0.6, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'우유'}), confidence=0.6, lift=1.0)]) 

RelationRecord(items=frozenset({'계란', '우유'}), support=0.4, ordered_statistics=[OrderedStatistic(items_base=frozenset({'계란'}), items_add=frozenset({'우유'}), confidence=0.6666666666666667, lift=1.1111111111111114), OrderedStatistic(items_base=frozenset({'우유'}), items_add=frozenset({'계란'}), confidence=0.6666666666666667, lift=1.1111111111111114)]) 

RelationRecord(items=frozenset({'커피', '계란'}), support=0.4, ordered_statistics=[OrderedStatistic(items_base=frozenset({'계란'}), items_add=frozenset({'커피'}), confidence=0.6666666666666667, lift=1.6666666666666667), OrderedStatistic(items_base=frozenset({'커피'}), items_add=frozenset({'계란'}), confidence=1.0, li

##### 여러개의 연관규칙 중 세번째만 본다 (인덱스 번호가 2니까 세번째)
##### 빈발항목집합의 길이가 1인건 제외해야됨. (연관규칙 만들 수 없으니까)

In [13]:
r = list(rules[2]) # rules의 2번째 {계란, 우유}를 r에 넣어준다.

print("빈발항목집합 : ", r[0]) # 0번 인덱스는 항목집합을 보여주는 부분
print("지지도: ", r[1]) # 1번 인덱스는 지지도
print("연관규칙 정보 : ", r[2]) # 연관규칙이 뭔지, 신뢰도, 향상도가 모여있는 2
print("연관규칙의 개수 :", len(r[2]))

빈발항목집합 :  frozenset({'계란', '우유'})
지지도:  0.4
연관규칙 정보 :  [OrderedStatistic(items_base=frozenset({'계란'}), items_add=frozenset({'우유'}), confidence=0.6666666666666667, lift=1.1111111111111114), OrderedStatistic(items_base=frozenset({'우유'}), items_add=frozenset({'계란'}), confidence=0.6666666666666667, lift=1.1111111111111114)]
연관규칙의 개수 : 2


In [15]:
s = list(rules[2][2]) # 세번째의 연관규칙정보 
# 한개의 항목집합에는 여러개의 연관규칙 있을 수 있따. 때문에 s에는 여러 연관규칙 있을 수 있음

for i in s:
    print(i, "\n")

OrderedStatistic(items_base=frozenset({'계란'}), items_add=frozenset({'우유'}), confidence=0.6666666666666667, lift=1.1111111111111114) 

OrderedStatistic(items_base=frozenset({'우유'}), items_add=frozenset({'계란'}), confidence=0.6666666666666667, lift=1.1111111111111114) 



#### 아이템 베이스 뒤에는 선행절 
#### 아이템 애드 뒤에는 후행절 
#### 연관규칙의 모양은 계란 -> 우유 가 된다. 
#### 이 연관규칙의 신뢰도는 0.667
#### 이 연관규칙의 향상도는 1.1111 
### 계란, 우유 라는 빈발항목집합(아이템즈)에서 나온 결과이다

#### 우유 -> 계란 연관규칙 
#### 지금 보고 있는 데이터를 꾸며서 볼 수 있으면 좋을 것 같다.

In [16]:
p = list(rules[2][2][0])

print("선행절: ", p[0])
print("후행절 : ", p[1])
print("신뢰도: ", p[2])
print("향상도: ", p[3])

선행절:  frozenset({'계란'})
후행절 :  frozenset({'우유'})
신뢰도:  0.6666666666666667
향상도:  1.1111111111111114


In [17]:
for r in rules:
    if len(r[0]) > 1: # 항목집합의 길이가 1보다 큰것만 골라라 
        support = r[1]
        for i in list(r[2]): # 연관규칙 덩어리가 r[2]
            print("연관규칙 :", set(i[0]), "==>", set(i[1])) # 선행, 후행 가져와서 집합으로 만들어라 (대괄호 생기게)
            print("지지도 : ",round(support, 4))
            print("신뢰도 : ", round(i[2], 4))
            print("향상도 : ", round(i[3], 4))
            print()

연관규칙 : {'계란'} ==> {'우유'}
지지도 :  0.4
신뢰도 :  0.6667
향상도 :  1.1111

연관규칙 : {'우유'} ==> {'계란'}
지지도 :  0.4
신뢰도 :  0.6667
향상도 :  1.1111

연관규칙 : {'계란'} ==> {'커피'}
지지도 :  0.4
신뢰도 :  0.6667
향상도 :  1.6667

연관규칙 : {'커피'} ==> {'계란'}
지지도 :  0.4
신뢰도 :  1.0
향상도 :  1.6667

연관규칙 : {'빵'} ==> {'우유'}
지지도 :  0.4
신뢰도 :  1.0
향상도 :  1.6667

연관규칙 : {'우유'} ==> {'빵'}
지지도 :  0.4
신뢰도 :  0.6667
향상도 :  1.6667



##### 향상도가 높은 것부터 검토한다던지... 

In [18]:
import random

In [21]:
#dir(random) # 모듈의 함수, 변수 이름 (밑줄 아닌것만 함수)
help(random.sample)

Help on method sample in module random:

sample(population, k) method of random.Random instance
    Chooses k unique random elements from a population sequence or set.
    
    Returns a new list containing elements from the population while
    leaving the original population unchanged.  The resulting list is
    in selection order so that all sub-slices will also be valid random
    samples.  This allows raffle winners (the sample) to be partitioned
    into grand prize and second place winners (the subslices).
    
    Members of the population need not be hashable or unique.  If the
    population contains repeats, then each occurrence is a possible
    selection in the sample.
    
    To choose a sample in a range of integers, use range as an argument.
    This is especially fast and space efficient for sampling from a
    large population:   sample(range(10000000), 60)



In [25]:
random.sample(range(1, 46), 6)

[21, 20, 5, 8, 41, 19]

## 샘플을 간편하게 50개 뽑기

In [31]:
# 트랜잭션의 개수 총 수 
tr_num = 50

# 한 트랜잭션 내의 항목의 개수  (범위)
tr_len = list(range(2, 6))

# 항목이 가질 수 있는 값의 범위 
item_domain = list(range(1,21)) # 상품아이디가 1부터 20까지  

In [33]:
item_domain

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]

In [34]:
data = [] # 얘 리스트, 트랜잭션도 리스트니까 트랜잭션 넣으면 이차원 됨

for a in range(tr_num):
    c = random.sample(item_domain, random.choice(tr_len)) # 범위, 개수
    c.sort()
    data.append(c)

In [35]:
data

[[4, 7, 9],
 [1, 4, 13, 15],
 [1, 3, 4, 8, 12],
 [6, 13],
 [8, 9, 11],
 [8, 10, 12, 13],
 [1, 4, 6, 9, 17],
 [1, 11, 14, 15, 19],
 [2, 4, 7, 17],
 [5, 6, 7, 8, 12],
 [3, 10, 16, 20],
 [2, 10, 11, 14, 20],
 [1, 3, 20],
 [1, 18, 20],
 [7, 11, 16, 17],
 [5, 7, 9],
 [1, 3, 6, 15],
 [6, 7, 14],
 [6, 12, 20],
 [4, 12, 18, 20],
 [9, 12, 14, 15],
 [1, 14, 18],
 [1, 2, 4, 14],
 [4, 9, 11, 15, 20],
 [1, 3, 17, 18],
 [4, 5, 13, 15],
 [1, 4, 7],
 [6, 15, 18],
 [2, 3, 14, 16, 19],
 [12, 18],
 [3, 10],
 [3, 10, 16],
 [16, 17],
 [5, 6, 11, 15, 17],
 [2, 4, 7],
 [3, 10, 11, 20],
 [1, 5],
 [2, 19, 20],
 [5, 7, 13, 16, 17],
 [2, 10, 11, 16, 17],
 [8, 11, 12, 18],
 [5, 9, 12, 17, 18],
 [4, 6, 9],
 [4, 7],
 [1, 6, 8, 19],
 [3, 9, 13, 17],
 [11, 14, 16, 17, 19],
 [10, 15, 17, 18],
 [2, 10, 11],
 [1, 5, 7, 9, 16]]

# 연관규칙 탐사

In [37]:
import apyori as ap

rules = list(ap.apriori(data, min_support = 0.1, min_confidence = 0.2)) # 최소신뢰도는 보통 최소지지도보다 좀 높게 한다.

#### 리스트 변환 꼭 해야한다. 

In [38]:
rules # 너무 안보이니까 잘 보이게 바꾸자

[RelationRecord(items=frozenset({1}), support=0.28, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({1}), confidence=0.28, lift=1.0)]),
 RelationRecord(items=frozenset({3}), support=0.2, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({3}), confidence=0.2, lift=1.0)]),
 RelationRecord(items=frozenset({4}), support=0.26, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({4}), confidence=0.26, lift=1.0)]),
 RelationRecord(items=frozenset({6}), support=0.2, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({6}), confidence=0.2, lift=1.0)]),
 RelationRecord(items=frozenset({7}), support=0.22, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({7}), confidence=0.22, lift=1.0)]),
 RelationRecord(items=frozenset({9}), support=0.2, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({9}), confidence=0.2, lift=

In [42]:
for r in rules:
    if len(r[0]) > 1:
        support = r[1]
        ar = list(r[2])
        for a in ar:
            print(set(a[0]), " ==>", set(a[1]))
            print("support = ", r[1])
            print('confidence = ', a[2])
            print("lift = ", a[3])
            print()

{1}  ==> {4}
support =  0.1
confidence =  0.35714285714285715
lift =  1.3736263736263736

{4}  ==> {1}
support =  0.1
confidence =  0.38461538461538464
lift =  1.3736263736263736

{4}  ==> {7}
support =  0.1
confidence =  0.38461538461538464
lift =  1.7482517482517483

{7}  ==> {4}
support =  0.1
confidence =  0.4545454545454546
lift =  1.7482517482517483

{16}  ==> {17}
support =  0.1
confidence =  0.5555555555555556
lift =  2.314814814814815

{17}  ==> {16}
support =  0.1
confidence =  0.4166666666666667
lift =  2.314814814814815



In [49]:
import pandas as pd 

result = []

for r in rules:
    if len(r[0]) > 1 :
        support = r[1]
        ar = list(r[2]) #연관규칙 덩어리
        for a in ar:
            row = [set(a[0]), "==>", set(a[1]), support, round(a[2],4), round(a[3], 4)] # 한 행이다. 
            result.append(row)

col = ["선행절", "==>", "후행절", "지지도","신뢰도","향상도"]
df = pd.DataFrame(result, columns=col, index = range(1, len(result)+1))

In [50]:
df

선행절  ==>   후행절  지지도     신뢰도     향상도
1   {1}  ==>   {4}  0.1  0.3571  1.3736
2   {4}  ==>   {1}  0.1  0.3846  1.3736
3   {4}  ==>   {7}  0.1  0.3846  1.7483
4   {7}  ==>   {4}  0.1  0.4545  1.7483
5  {16}  ==>  {17}  0.1  0.5556  2.3148
6  {17}  ==>  {16}  0.1  0.4167  2.3148